src/core/EdFi.DataManagementService.Core/Middleware/RequestDataBodyLoggingMiddleware.cs

```diff
@@ -0,0 +1,100 @@
+// SPDX-License-Identifier: Apache-2.0
+// Licensed to the Ed-Fi Alliance under one or more agreements.
+// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+// See the LICENSE and NOTICES files in the project root for more information.
+
+using System.Text.Json;
+using System.Text.RegularExpressions;
+using EdFi.DataManagementService.Core.Pipeline;
+using Microsoft.Extensions.Logging;
+
+namespace EdFi.DataManagementService.Core.Middleware;
+
+internal class RequestDataBodyLoggingMiddleware(ILogger _logger, bool _maskRequestBodyInLogs) : IPipelineStep
+{
+    private const string MessageBody =
+        "Incoming {Method} request to {Path} with body structure: {Body} - {TraceId}";
+
+    public async Task Execute(PipelineContext context, Func<Task> next)
+    {
+        if (_logger.IsEnabled(LogLevel.Debug) && !string.IsNullOrEmpty(context.FrontendRequest.Body))
+        {
+            _logger.LogDebug(
+                "Entering RequestDataBodyLoggingMiddleware - {TraceId}",
+                context.FrontendRequest.TraceId
+            );
+
+            string body = SanitizeInput(context.FrontendRequest.Body);
+
+            if (!_maskRequestBodyInLogs)
+            {
+                _logger.LogDebug(
+                    MessageBody,
+                    context.Method,
+                    context.FrontendRequest.Path,
+                    body,
+                    context.FrontendRequest.TraceId
+                );
+            }
+            else
+            {
+                _logger.LogDebug(
+                    MessageBody,
+                    context.Method,
+                    context.FrontendRequest.Path,
+                    MaskRequestBody(body, _logger),
+                    context.FrontendRequest.TraceId
+                );
+            }
+        }
+        await next();
+    }
+
+    private static string SanitizeInput(string input)
+    {
+        // Deletes new line, line feed, carriage return and tab characters
+        return Regex.Replace(
+            input.Replace(Environment.NewLine, "").Replace("\n", "").Replace("\r", "").Replace("\t", ""),
+            @"\s+",
+            " "
+        );
+    }
+
+    private static string MaskRequestBody(string body, ILogger logger)
+    {
+        try
+        {
+            // Deserialize the JSON body in a dynamic object
+            JsonDocument jsonDoc = JsonDocument.Parse(body);
+            JsonElement maskedJson = MaskJsonElement(jsonDoc.RootElement);
+            return JsonSerializer.Serialize(maskedJson);
+        }
+        catch (JsonException ex)
+        {
+            logger.LogError(ex, "Error while masking request body.");
+            return body; // In case of error, it returns the original body.
```

Comment:
Let's return an empty string so that there is no chance of accidentally logging PII.